fabric-model-loading-api-v1/src/client/java/net/fabricmc/fabric/impl/client/model/loading/CompositeBlockStateModel.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.client.model.loading;
+
+import java.util.List;
+import java.util.function.Predicate;
+
+import com.mojang.serialization.MapCodec;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.block.BlockState;
+import net.minecraft.client.render.model.Baker;
+import net.minecraft.client.render.model.BlockModelPart;
+import net.minecraft.client.render.model.BlockStateModel;
+import net.minecraft.client.texture.Sprite;
+import net.minecraft.util.dynamic.Codecs;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+import net.minecraft.util.math.random.Random;
+import net.minecraft.world.BlockRenderView;
+
+import net.fabricmc.fabric.api.client.model.loading.v1.CustomUnbakedBlockStateModel;
+import net.fabricmc.fabric.api.renderer.v1.mesh.QuadEmitter;
+
+public class CompositeBlockStateModel implements BlockStateModel {
+	private final BlockStateModel[] models;
+
+	public CompositeBlockStateModel(BlockStateModel[] models) {
+		this.models = models;
+	}
+
+	@Override
+	public void addParts(Random random, List<BlockModelPart> parts) {
+		long seed = random.nextLong();
+
+		for (BlockStateModel model : models) {
+			random.setSeed(seed);
+			model.addParts(random, parts);
+		}
+	}
+
+	@Override
+	public void emitQuads(QuadEmitter emitter, BlockRenderView blockView, BlockPos pos, BlockState state, Random random, Predicate<@Nullable Direction> cullTest) {
+		long seed = random.nextLong();
+
+		for (BlockStateModel model : models) {
+			random.setSeed(seed);
+			model.emitQuads(emitter, blockView, pos, state, random, cullTest);
+		}
+	}
+
+	@Override
+	@Nullable
+	public Object createGeometryKey(BlockRenderView blockView, BlockPos pos, BlockState state, Random random) {
+		int count = models.length;
+		long seed = random.nextLong();
+
+		if (count == 1) {
+			random.setSeed(seed);
+			return models[0].createGeometryKey(blockView, pos, state, random);
+		} else {
+			Object[] subkeys = new Object[count];
+
+			for (int i = 0; i < count; i++) {
+				random.setSeed(seed);
+				Object subkey = models[i].createGeometryKey(blockView, pos, state, random);
+
+				if (subkey == null) {
+					return null;
+				}
+
+				subkeys[i] = subkey;
+			}
+
+			record Key(Object[] subkeys) {
```

Comment:
Are you sure that this will produce suitable `equals` and `hashCode`?

Reply:
I don't see why not; it is almost exactly the same code as is used by the multipart model.

Reply:
It turns out arrays do not override `equals` and `hashCode` for whatever reason. I will replace the array with a `List`.